File: ios/Sources/NetworkDiscoveryPlugin/NetworkDiscoveryPlugin.swift
import Foundation
import Capacitor

@objc(NetworkDiscoveryPlugin)
public class NetworkDiscoveryPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NetworkDiscoveryPlugin"
    public let jsName = "NetworkDiscovery"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "startServer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopServer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "findServer", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = NetworkDiscovery()

    @objc func startServer(_ call: CAPPluginCall) {
        guard let name = call.getString("serviceName"),
              let type = call.getString("serviceType"),
              let port = call.getInt("port") else { return call.reject("Faltan parámetros") }
        
        let metadata = call.getObject("metadata") as? [String: String] ?? [:]
        do {
            try implementation.startPublishing(name: name, type: type, port: port, metadata: metadata)
            call.resolve()
        } catch { call.reject(error.localizedDescription) }
    }

    @objc func stopServer(_ call: CAPPluginCall) {
        implementation.stopServer()
        call.resolve()
    }

    @objc func findServer(_ call: CAPPluginCall) {
        guard let name = call.getString("serviceName"),
              let type = call.getString("serviceType") else { return call.reject("Faltan parámetros") }
        
        let timeout = Double(call.getInt("timeout") ?? 10000) / 1000.0
        implementation.findService(name: name, type: type, timeout: timeout) { result in
            if let data = result { call.resolve(data) } 
            else { call.reject("TIMEOUT_ERROR") }
        }
    }
}

File: android/src/main/java/com/cappitolian/plugins/networkdiscovery/NetworkDiscoveryPlugin.java
package com.cappitolian.plugins.networkdiscovery;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

@CapacitorPlugin(name = "NetworkDiscovery")
public class NetworkDiscoveryPlugin extends Plugin {
    private NetworkDiscovery implementation;

    @Override
    public void load() {
        implementation = new NetworkDiscovery(getContext());
    }

    @PluginMethod
    public void startServer(PluginCall call) {
        String name = call.getString("serviceName");
        String type = call.getString("serviceType");
        Integer port = call.getInt("port", 8081);
        JSObject metaJS = call.getObject("metadata");

        Map<String, String> metadata = new HashMap<>();
        if (metaJS != null) {
            Iterator<String> keys = metaJS.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                metadata.put(key, metaJS.getString(key));
            }
        }

        implementation.startServer(name, type, port, metadata, new NetworkDiscovery.Callback() {
            @Override public void success(JSObject data) { call.resolve(data); }
            @Override public void error(String msg) { call.reject(msg); }
        });
    }

    @PluginMethod
    public void stopServer(PluginCall call) {
        implementation.stopServer(new NetworkDiscovery.Callback() {
            @Override public void success(JSObject data) { call.resolve(data); }
            @Override public void error(String msg) { call.reject(msg); }
        });
    }

    @PluginMethod
    public void findServer(PluginCall call) {
        String name = call.getString("serviceName");
        String type = call.getString("serviceType");
        Integer timeout = call.getInt("timeout", 10000);

        implementation.findServer(name, type, timeout, new NetworkDiscovery.Callback() {
            @Override public void success(JSObject data) { call.resolve(data); }
            @Override public void error(String msg) { call.reject(msg); }
        });
    }
}